### PR TITLE
Scan track designs from multiple locations and refactor

### PR DESCRIPTION
**Pull request by IntelOrca**
Originally intended to allow track designs to be loaded from `%RCT2%/Tracks` as well as `%USER%/tracks` recursively, but ended up also being a major refactor and clean up of track designs.

A lot of the track design code was in track, so I have split that up and also separated index and saving code into their own sources. There is still a of a lot of code related to track designs and further cleanup is definitely required, but I don't feel like doing any more right now.

I broke many things doing this and then fixed them all, but its likely there are still some bugs which I have introduced so this needs more testing. The following things should be tested:

* Track preview from in game and in track manager
* Track saving, with and without scenery
* Track installing from track manager
* Track rename and delete from track manager
* Track organisation in tracks directories etc.

x2 as tracked rides and mazes work in completely different ways annoyingly.

@Gymnasiast I saw your comment about ```byte_F4414E```, but I have decided to leave it for now as I don't know what all the flags in it mean.

Could someone also please PR an xcode project fix, thanks.

### CODE REVIEW

File: src/localisation/string_ids.h [hunk 1]
```diff
@@ -2443,13 +2450,16 @@ enum {
 	STR_SHORTCUT_MUTE_SOUND = 5804,
 	
 	STR_ADVERTISE_SERVER_TIP = 5805,
-
+
```
Comment (IntelOrca) [inline src/localisation/string_ids.h]: Remove this

File: openrct2.vcxproj.filters [hunk 2]
```diff
@@ -604,6 +604,16 @@
       <Filter>Source\CommandLine</Filter>
     </ClCompile>
     <ClCompile Include="src\interface\paint_surface.c" />
+    <ClCompile Include="src\ride\track_list.c" />
```
Comment (IntelOrca) [inline openrct2.vcxproj.filters]: Remove this

**Comment by janisozaur** [comment]
```
/home/janisozaur/workspace/OpenRCT2/src/windows/track_list.c:202:31: warning: comparison of constant -1 with expression of type 'uint16' (aka 'unsigned short') is always true
      [-Wtautological-constant-out-of-range-compare]
                if (_loadedTrackDesignIndex != -1 && (_loadedTrackDesign->track_flags & 4)) {
                    ~~~~~~~~~~~~~~~~~~~~~~~ ^  ~~
/home/janisozaur/workspace/OpenRCT2/src/windows/track_list.c:395:10: warning: initializing 'uint8 *' (aka 'unsigned char *') with an expression of type 'utf8 *' (aka 'char *') converts between pointers to
      integer types with different sign [-Wpointer-sign]
                uint8 *path = _trackDesigns[trackIndex].path;
                       ^      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/janisozaur/workspace/OpenRCT2/src/windows/track_list.c:396:42: warning: passing 'uint8 *' (aka 'unsigned char *') to parameter of type 'utf8 *' (aka 'char *') converts between pointers to integer
      types with different sign [-Wpointer-sign]
                if (track_list_load_design_for_preview(path)) {
                                                       ^~~~
/home/janisozaur/workspace/OpenRCT2/src/windows/track_list.c:109:54: note: passing argument to parameter 'path' here
static bool track_list_load_design_for_preview(utf8 *path);
                                                     ^
3 warnings generated.
```

```
/home/janisozaur/workspace/OpenRCT2/src/ride/track_design_save.c:43:43: warning: 'memset' call operates on objects of type 'rct_map_element *' while the size is based on a different type
      'rct_map_element **' [-Wsizeof-pointer-memaccess]
        memset(_trackSavedMapElements, 0, sizeof(_trackSavedMapElements));
               ~~~~~~~~~~~~~~~~~~~~~~            ^~~~~~~~~~~~~~~~~~~~~~
/home/janisozaur/workspace/OpenRCT2/src/ride/track_design_save.c:43:43: note: did you mean to dereference the argument to 'sizeof' (and multiply it by the number of elements)?
        memset(_trackSavedMapElements, 0, sizeof(_trackSavedMapElements));
                                                 ^~~~~~~~~~~~~~~~~~~~~~
1 warning generated.
```

File: data/language/en-GB.txt [hunk 3]
```diff
@@ -4114,8 +4114,9 @@ STR_5805    :{SMALLFONT}{BLACK}If checked, your server will be added to the{NEWL
 STR_5807    :{WINDOW_COLOUR_2}Number of rides: {BLACK}{COMMA16}
 STR_5808    :{WINDOW_COLOUR_2}Number of shops and stalls: {BLACK}{COMMA16}
 STR_5809    :{WINDOW_COLOUR_2}Number of information kiosks and other facilities: {BLACK}{COMMA16}
-
 STR_5810    :Disable train length limit
+STR_5811    :{OPENQUOTES}{STRING}{ENDQUOTES}
```
Comment (Nubbie) [inline data/language/en-GB.txt]: The empty line on 4117 has been removed and the STR_5811 are now in use from PR #3480

File: src/localisation/string_ids.h [hunk 4]
```diff
@@ -2443,13 +2450,16 @@ enum {
...
 	STR_ADVERTISE_SERVER_TIP = 5805,
-
+	
 	STR_NUMBER_RIDES = 5807,
 	STR_NUMBER_SHOPS_AND_STALLS = 5808,
 	STR_NUMBER_RESTROOMS_AND_INFORMATION_KIOSKS = 5809,
 
 	STR_CHEAT_DISABLE_TRAIN_LENGTH_LIMIT = 5810,
 
+	STR_TRACK_LIST_NAME_FORMAT = 5811,
```
Comment (Nubbie) [inline src/localisation/string_ids.h]: STR_5811 are now in use from PR #3480

**Comment by IntelOrca** [comment]
Has anyone tested this yet?

Also could someone please PR an update to the xcode project, thanks.

**Comment by YJSoft** [comment]
No track design at OSX.


(at track design manager)


(at in game - track design window skipped since there's no design)


...but there are design files at my RCT2 folder.

adding it manually not works too.


(first attempt)


(second attempt - this time, game asks other name)


(td6 file is saved, but not loaded)

**Comment by YJSoft** [comment]
https://github.com/OpenRCT2/OpenRCT2/blob/develop/src/platform/posix.c#L600 is returning 0(means error) and error is `No such file or directory`

**Comment by IntelOrca** [comment]
@YJSoft tracks2.idx is always created on startup, so I don't think your Windows test proves much, unless you swapped it after starting OpenRCT2. Would the paths stored in tracks2.idx even work though if it was generated on OSX?

**Comment by YJSoft** [comment]
@IntelOrca 
Windows - `54 49 44 58 00 00 00 00`
OS X - `54 49 44 58 00 00 00 F3`

changing `F3` to `00` solves problem.

**Comment by IntelOrca** [comment]
Ah, thanks I know what's wrong now, I'll fix it later.

**Comment by IntelOrca** [comment]
@YJSoft I have fixed that issue now, could you try it again?

**Comment by Gymnasiast** [comment]
Some things I noticed (Ubuntu 16.04):
- It always uses the native dialog, rather than our own.
- It sets the title of that dialog to `[000E]29°C`.
- If I type a name without an extension, none is appended.
- It crashes when I try to place back a track design I just saved. (A steel twister with a monorail train, but still that shouldn't crash.)
- It tries to save to the old location (not just confined to the track save dialog, by the way)

**Comment by IntelOrca** [comment]
> It sets the title of that dialog to [000E]29°C.
> If I type a name without an extension, none is appended.
> It tries to save to the old location (not just confined to the track save dialog, by the way)

I can't reproduce these suggesting there is an issue with the Linux native file dialog API.

> It crashes when I try to place back a track design I just saved. (A steel twister with a monorail train, but still that shouldn't crash.)

Could you send me the TD6 please.

**Comment by janisozaur** [comment]
I tried installing a track design with TDM, `Batty Bikers.TD6`, and it caused this assertion to fail:
```diff
diff --git a/src/util/sawyercoding.c b/src/util/sawyercoding.c
index 68abb77..e0c592c 100644
--- a/src/util/sawyercoding.c
+++ b/src/util/sawyercoding.c
@@ -251,6 +251,7 @@ size_t sawyercoding_encode_td6(const uint8* src, uint8* dst, size_t length){
 
 /* Based off of rct2: 0x006770C1 */
 int sawyercoding_validate_track_checksum(const uint8* src, size_t length){
+       assert(length > 4);
        uint32 file_checksum = *((uint32*)&src[length - 4]);
 
        uint32 checksum = 0;
```

**Comment by YJSoft** [comment]
@IntelOrca In OSX 10.11.4, Everything works fine except
- Cannot load scenery(I even tried to load at same park!)
- Ride enterance/exit not saved(It was saved before even without scenery)

**Comment by janisozaur** [comment]
When selecting "Save Track Design", the dialog's title is set to "Save Track Design with Scenery"

zaznaczenie_118

**Comment by IntelOrca** [comment]
> > If I type a name without an extension, none is appended.
> +1 on that

Please implement this in your Linux file dialog API like I did for Windows.

**Comment by janisozaur** [comment]
It seems the dialog title bug is perhaps a case of the title data being overwritten at some point before the dialog is shown. Not the dialog title and latest message in following screenshot:

zaznaczenie_119

**Comment by janisozaur** [comment]
The source of invalid dialog title is here:
https://github.com/IntelOrca/OpenRCT2/blob/feature/multiple-trackdesign-locations/src/ride/track_design_save.c#L182

Seems like unparsed string? Can you fix @IntelOrca ?

**Comment by janisozaur** [comment]
* `tracks` folder is not created if it doesn't exist.
* Invalid title (see previous comment)
* `bh = bh;` https://github.com/IntelOrca/OpenRCT2/blob/feature/multiple-trackdesign-locations/src/ride/track_design.c#L638

^ I think these should still get solved for this to go in.

* If I type a name without an extension, none is appended.

^ This is exposed by, but not otherwise connected to this PR and can be solved on its own.

**Comment by janisozaur** [comment]
@YJSoft can you check if the issues you reported are solved now?

**Pull request merged**